expression/constant_propagation.go

```diff
@@ -0,0 +1,235 @@
+				rID := s.getColID(rCol)
+				s.transitiveMatrix[lID][rID] = true
+				s.transitiveMatrix[rID][lID] = true
+			}
+		}
+	}
+	colLen := len(s.colMapper)
+	// We implement a floyd-warshall algorithm, see https://en.wikipedia.org/wiki/Floyd%E2%80%93Warshall_algorithm.
+	for k := 0; k < colLen; k++ {
+		for i := 0; i < colLen; i++ {
+			for j := 0; j < colLen; j++ {
+				if !s.transitiveMatrix[i][j] {
+					s.transitiveMatrix[i][j] = s.transitiveMatrix[i][k] && s.transitiveMatrix[k][j]
+				}
+			}
+		}
+	}
+	condsLen := len(s.conditions)
+	for i := 0; i < condsLen; i++ {
+		cond := s.conditions[i]
+		col, con := s.validPropagateCond(cond, inEqFuncNameMap)
+		if col != nil {
+			id := s.getColID(col)
+			for to, connected := range s.transitiveMatrix[id] {
+				if to != id && connected {
+					newFunc, _ := NewFunction(cond.(*ScalarFunction).FuncName.L, cond.GetType(), s.columns[to], con)
+					s.conditions = append(s.conditions, newFunc)
+				}
+			}
+		}
+	}
+}
+
+// propagatesEQ propagates equal expression multiple times. Like a = d and b * 2 = c and c = d + 2 and b = 1, the process is:
+// a = d & b * 2 = c & c = d + 2 & b = 1 & a = 4, we pick eq cond b = 1 and a = 4
+// d = 4 & 2 = c & c = d + 2 & b = 1 & a = 4, we propagate b = 1 and a = 4 and pick eq cond c = 2 and d = 4
+// d = 4 & 2 = c & false & b = 1 & a = 4, we propagate c = 2 and d = 4, and do constant folding: c = d + 2 will be folded as false.
+func (s *propagateConstantSolver) propagateEQ() {
+	s.eqList = make([]*Constant, len(s.columns))
+	visited := make([]bool, len(s.conditions))
+	for i := 0; i < MaxPropagateColsCnt; i++ {
+		mapper := s.pickNewEQConds(visited)
+		if s.foreverFalse || len(mapper) == 0 {
+			return
+		}
+		cols := make(Schema, 0, len(mapper))
+		cons := make([]Expression, 0, len(mapper))
+		for id, con := range mapper {
+			cols = append(cols, s.columns[id])
+			cons = append(cons, con)
+		}
+		for i, cond := range s.conditions {
+			if !visited[i] {
+				s.conditions[i] = ColumnSubstitute(cond, Schema(cols), cons)
+			}
+		}
+	}
+}
+
+// In this function we check if the cond is an expression like [column op constant] and op is in the funNameMap.
+func (s *propagateConstantSolver) validPropagateCond(cond Expression, funNameMap map[string]bool) (*Column, *Constant) {
+	if eq, ok := cond.(*ScalarFunction); ok {
+		if _, ok := funNameMap[eq.FuncName.L]; !ok {
+			return nil, nil
+		}
+		if col, colOk := eq.Args[0].(*Column); colOk {
+			if con, conOk := eq.Args[1].(*Constant); conOk {
+				return col, con
+			}
+		}
+		if col, colOk := eq.Args[1].(*Column); colOk {
+			if con, conOk := eq.Args[0].(*Constant); conOk {
+				return col, con
+			}
+		}
+	}
+	return nil, nil
+}
+
+// pickNewEQConds tries to pick new equal conds and puts them to retMapper.
+func (s *propagateConstantSolver) pickNewEQConds(visited []bool) (retMapper map[int]*Constant) {
+	retMapper = make(map[int]*Constant)
+	for i, cond := range s.conditions {
+		if !visited[i] {
```

Comment:
if visited[i] {
    continue
}

this can reduce the layers, may be more elegant.